src/node/stage/approved.rs

```diff
@@ -231,16 +231,55 @@ impl Approved {
     }
 
     pub fn handle_timeout(&mut self, core: &mut Core, token: u64) {
-        if self.timer_token == token {
+        if self.gossip_timer_token == token {
             if self.is_our_elder(core.id()) {
-                self.timer_token = core.timer.schedule(self.consensus_engine.gossip_period());
+                self.gossip_timer_token =
+                    core.timer.schedule(self.consensus_engine.gossip_period());
                 self.consensus_engine.reset_gossip_period();
             } else {
                 // TODO: send this only when the knowledge changes, not periodically.
                 self.send_parsec_poke(core);
-                self.timer_token = core.timer.schedule(KNOWLEDGE_TIMEOUT);
+                self.gossip_timer_token = core.timer.schedule(KNOWLEDGE_TIMEOUT);
+            }
+        } else if self.dkg_voter.timer_token() == token {
+            self.dkg_voter
+                .set_timer_token(core.timer.schedule(DKG_PROGRESS_INTERVAL));
+            self.progress_dkg(core);
+        }
+    }
+
+    fn check_dkg(&mut self, core: &mut Core) {
+        let (completed, mut backlog_events) = self.dkg_voter.check_dkg();
+
+        for (dkg_key, dkg_result) in completed {
+            debug!("Completed DKG {:?}", dkg_key);
+            self.notify_sibling(
+                core,
+                &dkg_key.0,
+                dkg_key.1,
+                dkg_result.public_key_set.clone(),
+            );
+            if let Err(err) = self.handle_dkg_result_event(core, &dkg_key.0, &dkg_result) {
+                debug!("Failed handle DKG result of {:?} - {:?}", dkg_key, err);
+            } else {
+                self.dkg_voter.remove_voter(&dkg_key);
             }
         }
+
+        // To avoid the case that DKG was completed after received certain accumulated events.
+        while let Some((event, proof)) = backlog_events.pop_back() {
+            trace!("handle cached accumulated event {:?}", event);
+            // In case of error, event got cached inside `handle_accumulated_event`.
+            let _ = self.handle_accumulated_event(core, event.clone(), Some(proof.clone()));
```

Comment:
Maybe safer would to only ignore the `InvalidDkgResult` error and still propagate the rest?

Reply:
it's actually handling a cached event. propagate the error will be tricky (handle_time_out function will be affected).
Actually, this cached event shall only cause `InvalidDkgResult` error, as this is why the event got cached.

Reply:
OK, then maybe at least log it here?

Reply:
Ok, will have it logged